### PR TITLE
KAFKA-9294: Add tests for Named parameter

### PR DESCRIPTION
**Pull request by mjsax**
Part (2) -- stateful KStream operators

Follow up to #7874  

KStream interface contains some method reordering (I also renamed some parameters as highlighted)

Call for review @bbejeck @vvcephei

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 1]
```diff
@@ -875,192 +875,192 @@ void to(final TopicNameExtractor<K, V> topicExtractor,
...
+     * This operation is equivalent to calling {@link #selectKey(KeyValueMapper)} followed by {@link #groupByKey()}.
+     * If the key type is changed, it is recommended to use {@link #groupBy(KeyValueMapper, Grouped)} instead.
      *
+     * @param keySelector a {@link KeyValueMapper} that computes a new key for grouping
+     * @param <KR>        the key type of the result {@link KGroupedStream}
      * @return a {@link KGroupedStream} that contains the grouped records of the original {@code KStream}
-     * @see #groupBy(KeyValueMapper)
      */
-    KGroupedStream<K, V> groupByKey();
+    <KR> KGroupedStream<KR, V> groupBy(final KeyValueMapper<? super K, ? super V, KR> keySelector);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:905]: renamed `selector` to `keySelector`

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 2]
```diff
@@ -875,192 +875,192 @@ void to(final TopicNameExtractor<K, V> topicExtractor,
...
+     * @param <KR>        the key type of the result {@link KGroupedStream}
      * @return a {@link KGroupedStream} that contains the grouped records of the original {@code KStream}
-     * @see #groupBy(KeyValueMapper)
      *
-     * @deprecated since 2.1. Use {@link org.apache.kafka.streams.kstream.KStream#groupByKey(Grouped)} instead
+     * @deprecated since 2.1. Use {@link org.apache.kafka.streams.kstream.KStream#groupBy(KeyValueMapper, Grouped)} instead
      */
     @Deprecated
-    KGroupedStream<K, V> groupByKey(final Serialized<K, V> serialized);
+    <KR> KGroupedStream<KR, V> groupBy(final KeyValueMapper<? super K, ? super V, KR> keySelector,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:937]: renamed `selector` to `keySelector`

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 3]
```diff
@@ -875,192 +875,192 @@ void to(final TopicNameExtractor<K, V> topicExtractor,
...
-     *                  and part of the name for a repartition topic if repartitioning is required.
+     * @param keySelector a {@link KeyValueMapper} that computes a new key for grouping
+     * @param grouped     the {@link Grouped} instance used to specify {@link org.apache.kafka.common.serialization.Serdes}
+     *                    and part of the name for a repartition topic if repartitioning is required.
+     * @param <KR>        the key type of the result {@link KGroupedStream}
      * @return a {@link KGroupedStream} that contains the grouped records of the original {@code KStream}
-     * @see #groupBy(KeyValueMapper)
      */
-    KGroupedStream<K, V> groupByKey(final Grouped<K, V> grouped);
+    <KR> KGroupedStream<KR, V> groupBy(final KeyValueMapper<? super K, ? super V, KR> keySelector,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:970]: renamed `selector` to `keySelector`

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 4]
```diff
@@ -2144,8 +2144,8 @@ void to(final TopicNameExtractor<K, V> topicExtractor,
      * {@link ValueJoiner}, one output for each input {@code KStream} record
      * @see #leftJoin(GlobalKTable, KeyValueMapper, ValueJoiner)
      */
-    <GK, GV, RV> KStream<K, RV> join(final GlobalKTable<GK, GV> globalKTable,
-                                     final KeyValueMapper<? super K, ? super V, ? extends GK> keyValueMapper,
+    <GK, GV, RV> KStream<K, RV> join(final GlobalKTable<GK, GV> globalTable,
+                                     final KeyValueMapper<? super K, ? super V, ? extends GK> keySelector,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:2148]: renamed `globalKTable` to `globalTable`
renamed `keyValueMapper ` to `keySelector`

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 5]
```diff
@@ -2178,8 +2178,8 @@ void to(final TopicNameExtractor<K, V> topicExtractor,
...
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:2182]: renamed `globalKTable` to `globalTable`
renamed `keyValueMapper ` to `keySelector`

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 6]
```diff
@@ -2216,8 +2216,8 @@ void to(final TopicNameExtractor<K, V> topicExtractor,
      * {@link ValueJoiner}, one output for each input {@code KStream} record
      * @see #join(GlobalKTable, KeyValueMapper, ValueJoiner)
      */
-    <GK, GV, RV> KStream<K, RV> leftJoin(final GlobalKTable<GK, GV> globalKTable,
-                                         final KeyValueMapper<? super K, ? super V, ? extends GK> keyValueMapper,
+    <GK, GV, RV> KStream<K, RV> leftJoin(final GlobalKTable<GK, GV> globalTable,
+                                         final KeyValueMapper<? super K, ? super V, ? extends GK> keySelector,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:2220]: renamed `globalKTable` to `globalTable`
renamed `keyValueMapper ` to `keySelector`

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 7]
```diff
@@ -2254,8 +2254,8 @@ void to(final TopicNameExtractor<K, V> topicExtractor,
...
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:2258]: renamed `globalKTable` to `globalTable`
renamed `keyValueMapper ` to `keySelector`

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 8]
```diff
@@ -601,30 +601,30 @@ private void to(final TopicNameExtractor<K, V> topicExtractor,
     }
 
     @Override
-    public <KR> KGroupedStream<KR, V> groupBy(final KeyValueMapper<? super K, ? super V, KR> selector) {
-        return groupBy(selector, Grouped.with(null, valSerde));
+    public <KR> KGroupedStream<KR, V> groupBy(final KeyValueMapper<? super K, ? super V, KR> keySelector) {
+        return groupBy(keySelector, Grouped.with(null, valSerde));
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:605]: As in the interface; renamed parameter

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 9]
```diff
@@ -664,10 +668,10 @@ private void to(final TopicNameExtractor<K, V> topicExtractor,
     }
 
     @Override
-    public <VO, VR> KStream<K, VR> join(final KStream<K, VO> other,
+    public <VO, VR> KStream<K, VR> join(final KStream<K, VO> otherStream,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:671]: It's called `otherStream` in the interface -- aligning both

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 10]
```diff
@@ -857,55 +900,16 @@ private void to(final TopicNameExtractor<K, V> topicExtractor,
...
-            other,
-            joiner,
-            windows,
-            streamJoined,
-            new KStreamImplJoin(builder, true, false));
-    }
-
-    @Override
-    public <VO, VR> KStream<K, VR> join(final KTable<K, VO> other,
+    public <VO, VR> KStream<K, VR> join(final KTable<K, VO> table,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:903]: It's called `table` in the interface -- aligning both

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 11]
```diff
@@ -946,49 +950,49 @@ private void to(final TopicNameExtractor<K, V> topicExtractor,
...
         } else {
-            return doStreamTableJoin(other, joiner, joined, true);
+            return doStreamTableJoin(table, joiner, joined, true);
         }
     }
 
     @Override
     public <KG, VG, VR> KStream<K, VR> join(final GlobalKTable<KG, VG> globalTable,
-                                            final KeyValueMapper<? super K, ? super V, ? extends KG> keyMapper,
+                                            final KeyValueMapper<? super K, ? super V, ? extends KG> keySelector,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:961]: As in the interface; renamed parameter

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 12]
```diff
@@ -946,49 +950,49 @@ private void to(final TopicNameExtractor<K, V> topicExtractor,
...
 
     private <KG, VG, VR> KStream<K, VR> globalTableJoin(final GlobalKTable<KG, VG> globalTable,
-                                                        final KeyValueMapper<? super K, ? super V, ? extends KG> keyMapper,
+                                                        final KeyValueMapper<? super K, ? super V, ? extends KG> keySelector,
                                                         final ValueJoiner<? super V, ? super VG, ? extends VR> joiner,
                                                         final boolean leftJoin,
                                                         final Named named) {
         Objects.requireNonNull(globalTable, "globalTable can't be null");
-        Objects.requireNonNull(keyMapper, "keyMapper can't be null");
+        Objects.requireNonNull(keySelector, "keySelector can't be null");
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:995]: As in the interface; renamed parameter
(note, that the first parameter is already called `globalTable`)

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamImplTest.java [hunk 13]
```diff
@@ -79,11 +79,14 @@
...
 import static org.junit.Assert.assertThrows;
 import static org.junit.Assert.assertTrue;
-import static org.junit.Assert.fail;
 
 public class KStreamImplTest {
 
     private final Consumed<String, String> stringConsumed = Consumed.with(Serdes.String(), Serdes.String());
+    @SuppressWarnings("deprecation")
+    private final org.apache.kafka.streams.kstream.Serialized<String, String> stringSerialized =
+        org.apache.kafka.streams.kstream.Serialized.with(Serdes.String(), Serdes.String());
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamImplTest.java:88]: Cannot use import as we cannot suppress deprecation warning on import (similar in the tests).

**Review by bbejeck — APPROVED** [review]
Thanks for cleanup LGTM

**Comment by bbejeck** [comment]
Java 11 failed with `kafka.admin.ReassignPartitionsClusterTest.shouldListMovingPartitionsThroughApi` Jira ticket updated
Java 8 passed

retest this please

**Comment by mjsax** [comment]
Java 8:
```
kafka.admin.ReassignPartitionsClusterTest.shouldExpandCluster
org.apache.kafka.streams.integration.BranchedMultiLevelRepartitionConnectedTopologyTest.testTopologyBuild
```
Java 11 failed with env issue.

Retest this please.

**Comment by mjsax** [comment]
Both builds failed -- test results are not available any longer.

Retest this please.

**Comment by mjsax** [comment]
Java 11:
```
kafka.api.GroupEndToEndAuthorizationTest.testNoDescribeProduceOrConsumeWithoutTopicDescribeAcl
```
Java8
```
org.apache.kafka.connect.mirror.MirrorConnectorsIntegrationTest.testReplication
```

This is different tests. Merging.

**Pull request merged**